### DISCUSSION TITLE
Allow direct import of JSON files

### DISCUSSION THREAD

**Discussion by derkoe** [ivanhofer/typesafe-i18n#151] [DX improvements]
Would be great to use JSON files directly as a source. This will make integrating translation tools easier.

This is best understood with an example:

`src/i18n/en/index.ts`:

```ts
import * as en from './texts.json';

export default en;
```

`src/i18n/en/texts.json`:
```json
{
    "HI": "Hallo {name}!",
    "BYE": "Bye bye"
}
```

Currently, when you do this you'll get the following error:

```
[typesafe-i18n] ERROR: import failed for .../node_modules/typesafe-i18n/temp-output/11/index.js Error: Cannot find module './texts.json'
Require stack:
- .../node_modules/typesafe-i18n/temp-output/11/index.js
[typesafe-i18n] ERROR: could not read default export from language file 'en'
```

**Comment by derkoe** [comment]
With this change the JSON imports would work:

https://github.com/ivanhofer/typesafe-i18n/compare/main...derkoe:allow-json

**Reply by ivanhofer** [reply]
Even if I dislike having `.json` files for the dictionaries, I will think about adding this because it's such a simple fix. I let you know once I know if I want to support such a use-case

**Reply by derkoe** [reply]
Here you go https://github.com/ivanhofer/typesafe-i18n/pull/153

**Comment by ivanhofer** [comment]
Hi @derkoe,
I have created this llibrary to get rid of error-prone `.json` files ;)
The problem with json files is that you can't really check if they contain valid content. Thats why you need to define the translations insode a file that supports static type checking. (`.ts` or `.js` with JSDoc annotations).

To support migrating from traditional i18n libraries and services, `typesafe-i18n` comes with an [`importer`](https://github.com/ivanhofer/typesafe-i18n#importer). You could use it to read the json file and convert it to a fully typesafe dictionary.
Let me know if you need some help :)

**Reply by derkoe** [reply]
The good thing about importing the JSON directly into TypeScript is that you get type errors:

```typescript
import type { Translation } from '../i18n-types'
import * as deJson from './de.json'
const de: Translation = deJson;
export default de
```
![image](https://user-images.githubusercontent.com/123199/149886987-66fc40f5-9b53-4931-bcfc-fd2c2b1c4b4c.png)

**Reply by ivanhofer** [reply]
Oh I didn't knew this works!
You won't see the errors inside the actual dictionary so you need to find the correct place where to fix it. This could be frustrating for larger dictionaries.
But it's better than nothing 😅

**Comment by ivanhofer** [comment]
support for (in-)direct imports of JSON files was added in version `2.59.0`
Thanks @derkoe for your help!

**Reply by bitdom8** [reply]
Does it allow json like with whitespace:

"hi sir":"hi sir",

Normally it is like
"HI_SIR": "Hi sir",

can we do this convention inside of TS files:

"hi sir":"hi sir",

I want translators to easily understand what's to be translated for german for instance. How do we allow translators to easily translate what's to be translated? It should be explicit. @ivanhofer 

const en: BaseTranslation = {
	"Hey how are you today?": "Hallo wie geht's ihnen heute",
}

Understood that this works. Tried already. Thanks
